### ISSUE TITLE
NSEC in correlation with CNAME returns wrong value

### ISSUE THREAD

**Issue by firstdorsal** [NLnetLabs/unbound#572]
**Describe the bug**

a query for an NSEC record returns the following when being sent to cloudflare

```sh
paul@firstdorsal ~> dig @1.1.1.1 autoconfig.y.gy NSEC

; <<>> DiG 9.16.22 <<>> @1.1.1.1 autoconfig.y.gy NSEC
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 18377
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;autoconfig.y.gy.		IN	NSEC

;; ANSWER SECTION:
autoconfig.y.gy.	3600	IN	NSEC	autodiscover.y.gy. CNAME RRSIG NSEC

;; Query time: 46 msec
;; SERVER: 1.1.1.1#53(1.1.1.1)
;; WHEN: Wed Nov 17 16:51:23 CET 2021
;; MSG SIZE  rcvd: 83
```

but returns this when sent through unbound
```sh
paul@firstdorsal ~> dig @localhost -p 5355 autoconfig.y.gy NSEC

; <<>> DiG 9.16.22 <<>> @localhost -p 5355 autoconfig.y.gy NSEC
; (2 servers found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37609
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;autoconfig.y.gy.		IN	NSEC

;; ANSWER SECTION:
autoconfig.y.gy.	10223	IN	CNAME	mx.y.gy.
mx.y.gy.		3023	IN	NSEC	_110._tcp.mx.y.gy. A AAAA RRSIG NSEC

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5355(127.0.0.1)
;; WHEN: Wed Nov 17 16:53:04 CET 2021
;; MSG SIZE  rcvd: 100

```

**To reproduce**
1. see above; tested with an unbound docker container on localhost and on a raspberry pi

**Expected behavior**

Should return the same as cloudflare google etc.

no idea what is rfc compliant in this case.

**System:**
 - Unbound version: 1.13.2
 - OS: arch/ubuntu
 - `unbound -V` output:
 ```sh
 Version 1.13.2

Configure line: --with-deprecate-rsa-1024 --disable-dependency-tracking --prefix=/opt/unbound --with-pthreads --with-username=_unbound --with-ssl=/opt/openssl --with-libevent --with-libnghttp2 --enable-tfo-server --enable-tfo-client --enable-event-api
Linked libs: libevent 2.1.12-stable (it uses epoll), OpenSSL 1.1.1k  25 Mar 2021
Linked modules: dns64 respip validator iterator
TCP Fastopen feature available

BSD licensed, see LICENSE in source package for details.
Report bugs to unbound-bugs@nlnetlabs.nl or https://github.com/NLnetLabs/unbound/issues
```

all the best and thank you for unbound!
paul

**Comment by wcawijngaards** [comment]
This happens because the upstream server responds with a CNAME. If unbound had hosted the authoritative zone, with eg. auth-zone, it would likely have responded with the NSEC response that you want. But the upstream authoritative server does not respond in that manner. It ignores the direct query for type NSEC and replies with the CNAME. Unbound can do nothing else but chase the cname.

You can see this with `dig @159.69.206.124 autoconfig.y.gy NSEC`
That responds with a CNAME for me.
```
$ dig @159.69.206.124 autoconfig.y.gy NSEC
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 32783
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;autoconfig.y.gy.		IN	NSEC

;; ANSWER SECTION:
autoconfig.y.gy.	10800	IN	CNAME	mx.y.gy.

;; AUTHORITY SECTION:
y.gy.			3600	IN	SOA	tick.y.gy. hostmaster.y.gy. 2021111303 10800 3600 604800 3600

;; Query time: 13 msec
;; SERVER: 159.69.206.124#53(159.69.206.124)
;; WHEN: Mon Nov 29 11:59:12 CET 2021
;; MSG SIZE  rcvd: 113
```

So the upstream server behaves in that manner. If Unbound was to query another type of upstream server that responds in a different manner, eg. with the correct NSEC record that you want, then you would get the response you are looking for.

**Comment by firstdorsal** [comment]
Thank you for the quick response.
I tinkered around for a bit, and it does not look like an upstream bug.

I am using PowerDNS, and it sure could be a bug in there, but other servers that use NSEC behave like this too (queried through unbound):
- Our local university: `autoconfig.hs-augsburg.de`
- One of the biggest German banks: `www.3dsecure.sparkasse.de`

Google, Cloudflare, OpenDns, Level3 give the expected, same answer when asked, and I could not figure out what they are doing to get it.

Verisign(64.6.64.6) gives exactly the same answer as unbound on my local machine. (maybe they are using unbound)

Note that Quad9(9.9.9.9) responds with NXDOMAIN

Further strange is that dig, when sending a query directly to the authoritative server like you did, also only gives back the CNAME and its resolution.

We thought that this part from the RFC might be related, but it couldn't help us finding the whole truth:
https://datatracker.ietf.org/doc/html/rfc4035#section-2.5
It explains how the public resolvers got to the DNSSEC types of the next record in the NSEC chain: 
`autoconfig.y.gy.	3600	IN	NSEC	autodiscover.y.gy. CNAME RRSIG NSEC` 
(CNAME and the signed zone implies the existence of `RRSIG` and `NSEC`) but not how they got  the TTL (3600) the next signed record (autodiscover.y.gy.) and the type of the next signed record (CNAME)

Thank you and have a great day.

**Comment by wcawijngaards** [comment]
Unbound gets the CNAME answer from the authoritative server, so it makes a CNAME response. I am closing the issue, because there seems to be nothing I can do, but agree that those differences exist in servers out there. But I have nothing to act on, so I am closing the issue. If you are looking for opinions, people to talk to, perhaps try the unbound-users list, or another discussion forum.

**Comment by firstdorsal** [comment]
Ok, I think I figured it out.

When adding the `+dnssec` option to dig, querying the authoritative server directly, it returns the expected record:

From the [dig manpage](https://man.openbsd.org/dig.1#+~16):

> +[no]dnssec
> Request DNSSEC records be sent by setting the DNSSEC OK bit (DO) in the OPT record in the additional section of the query (off by default).

```sh
paul@blackbox ~> dig NSEC autoconfig.y.gy. @2a01:4f8:c2c:3c85::1 +dnssec

; <<>> DiG 9.16.23 <<>> NSEC autoconfig.y.gy. @2a01:4f8:c2c:3c85::1 +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 64432
;; flags: qr aa rd; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1232
;; QUESTION SECTION:
;autoconfig.y.gy.		IN	NSEC

;; ANSWER SECTION:
autoconfig.y.gy.	3600	IN	RRSIG	NSEC 13 3 3600 20211209000000 20211118000000 38180 y.gy. 0XX4EF+a968lOBbiE8Wr4VLpnyayV0GLL+Llf0VrakIBawhpWKpcjwNz WwDCZwAauolUXLVVtCU17L6zJCT5bw==
autoconfig.y.gy.	3600	IN	NSEC	autodiscover.y.gy. CNAME RRSIG NSEC

;; Query time: 33 msec
;; SERVER: 2a01:4f8:c2c:3c85::1#53(2a01:4f8:c2c:3c85::1)
;; WHEN: Tue Nov 30 17:49:38 CET 2021
;; MSG SIZE  rcvd: 183
```

When querying **Unbound**, this is what happens:

```sh
paul@blackbox ~> dig NSEC autoconfig.y.gy. @localhost -p 5355 +dnssec

; <<>> DiG 9.16.23 <<>> NSEC autoconfig.y.gy. @localhost -p 5355 +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 45058
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1232
;; QUESTION SECTION:
;autoconfig.y.gy.		IN	NSEC

;; ANSWER SECTION:
autoconfig.y.gy.	10800	IN	CNAME	mx.y.gy.
autoconfig.y.gy.	10800	IN	RRSIG	CNAME 13 3 10800 20211209000000 20211118000000 38180 y.gy. Hu27hL6Iz8HCYIXz4bybLCP4P3P4hxXl1/K2aA025winILzuCtrxeUxg hJzH2PD3tOabl44Aqha62cwggXKYsA==
mx.y.gy.		3600	IN	NSEC	_110._tcp.mx.y.gy. A AAAA RRSIG NSEC
mx.y.gy.		3600	IN	RRSIG	NSEC 13 3 3600 20211209000000 20211118000000 38180 y.gy. gL712MR+kclfgtyj2S/rf2uzRle8lbOG/lNdpB4vtc1ljEVcjdj/xHUR xBs7nCbc4i/ONBgCm3rJSwU1Ul5Kkg==

;; Query time: 176 msec
;; SERVER: 127.0.0.1#5355(127.0.0.1)
;; WHEN: Tue Nov 30 17:49:56 CET 2021
;; MSG SIZE  rcvd: 300
```

This is **not** the expected behavior, as it should return an answer to the initial question.

This issue not only occurs when PowerDNS is the authoritative resolver (assuming nlnetlabs.nl is not using PowerDNS)

Asking the local Unbound resolver this is what comes up when querying for the NSEC record of `_acme-challenge.nlnetlabs.nl.` (a CNAME record)

```sh
paul@blackbox ~> dig NSEC _acme-challenge.nlnetlabs.nl. @localhost -p 5355

; <<>> DiG 9.16.23 <<>> NSEC _acme-challenge.nlnetlabs.nl. @localhost -p 5355
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 26550
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;_acme-challenge.nlnetlabs.nl.	IN	NSEC

;; ANSWER SECTION:
_acme-challenge.nlnetlabs.nl. 9815 IN	CNAME	nlnetlabs.nl.acme.nlnetlabs.nl.

;; AUTHORITY SECTION:
acme.nlnetlabs.nl.	3215	IN	SOA	ns.nlnetlabs.nl. sysadmin.nlnetlabs.nl. 1638247180 14400 3600 604800 3600

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5355(127.0.0.1)
;; WHEN: Tue Nov 30 18:16:44 CET 2021
;; MSG SIZE  rcvd: 137

```

and even with the `+dnssec` flag set

```sh
paul@blackbox ~> dig NSEC _acme-challenge.nlnetlabs.nl. @localhost -p 5355 +dnssec

; <<>> DiG 9.16.23 <<>> NSEC _acme-challenge.nlnetlabs.nl. @localhost -p 5355 +dnssec
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 8280
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 4, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1232
;; QUESTION SECTION:
;_acme-challenge.nlnetlabs.nl.	IN	NSEC

;; ANSWER SECTION:
_acme-challenge.nlnetlabs.nl. 9782 IN	CNAME	nlnetlabs.nl.acme.nlnetlabs.nl.
_acme-challenge.nlnetlabs.nl. 9782 IN	RRSIG	CNAME 8 3 10200 20211222015021 20211124015021 42393 nlnetlabs.nl. EYnzdU87i4bSf7jP3wxw9SdYPtdXrVCdZr2fvIaIk/pHWCaCq+5rqQ0r m0Op/qEGEtcEyGAiOxh7mZ05KqEFNsACkacUtvO7B52NNuvsE1tyW1cI 141ceiPvWePCleKv3ayuZqPpOk49/51EUKTudvEBfqD/IcGQeipyVkRp 7HY=

;; AUTHORITY SECTION:
hhb2ngapsbkfdpijh6nqq4qbepkp9sb3.acme.nlnetlabs.nl. 3182 IN NSEC3 1 0 5 34CCDEF2984D916D HHB2NGAPSBKFDPIJH6NQQ4QBEPKP9SB3 NS SOA TXT RRSIG DNSKEY NSEC3PARAM
hhb2ngapsbkfdpijh6nqq4qbepkp9sb3.acme.nlnetlabs.nl. 3182 IN RRSIG NSEC3 7 4 3600 20211221090130 20211130033940 35465 acme.nlnetlabs.nl. eTP+3XMq+fRJqUmYU09Qvm+q9+nJOzcN1+LmXzsuUEGJzi/FQqu8bI6R 5idzL4ZAtUWmQyVbtGzTGd9C3c0ipyXsEmvVu35SnIoEFoqgdU/VwvOl D1omCbfdMb9I+AFxxI2HHjiFmxGuOaOX2KJe1VWK11Ks3TTGYhGWRNzM H50=
acme.nlnetlabs.nl.	3182	IN	SOA	ns.nlnetlabs.nl. sysadmin.nlnetlabs.nl. 1638247180 14400 3600 604800 3600
acme.nlnetlabs.nl.	3182	IN	RRSIG	SOA 7 3 3600 20211220193001 20211130033940 35465 acme.nlnetlabs.nl. Dic9x0rhcBz4AQgIzwppt30jEi9guTWNOnykfjr4kDOGJwKvjABDMksD GdB49hFPU5TGS9cF7rHh1D1MD6DAxTt73U1Z7FSAyV8FNF2kv0S705FO DLsuGz1gV6OYr899sLjQu2/K9526yDq6eVFRWenwFcoDA+8l5TE8IPwd GIo=

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5355(127.0.0.1)
;; WHEN: Tue Nov 30 18:17:17 CET 2021
;; MSG SIZE  rcvd: 751
```

but when the query is sent directly to the authoritative server, ns.nlnetlabs.nl. dig gets the expected record (even without the `+dnssec` flag):

```sh
paul@blackbox ~> dig NSEC _acme-challenge.nlnetlabs.nl. @ns.nlnetlabs.nl

; <<>> DiG 9.16.23 <<>> NSEC _acme-challenge.nlnetlabs.nl. @ns.nlnetlabs.nl
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3527
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 3, ADDITIONAL: 3
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;_acme-challenge.nlnetlabs.nl.	IN	NSEC

;; ANSWER SECTION:
_acme-challenge.nlnetlabs.nl. 240 IN	NSEC	3.1.1._dane.nlnetlabs.nl. CNAME RRSIG NSEC

;; AUTHORITY SECTION:
nlnetlabs.nl.		240	IN	NS	ns.nlnetlabs.nl.
nlnetlabs.nl.		240	IN	NS	anyns.pch.net.
nlnetlabs.nl.		240	IN	NS	ns-ext1.sidn.nl.

;; ADDITIONAL SECTION:
ns.nlnetlabs.nl.	10200	IN	AAAA	2a04:b900::8:0:0:60
ns.nlnetlabs.nl.	10200	IN	A	185.49.140.60

;; Query time: 193 msec
;; SERVER: 2a04:b900::8:0:0:60#53(2a04:b900::8:0:0:60)
;; WHEN: Tue Nov 30 18:24:42 CET 2021
;; MSG SIZE  rcvd: 218

```

In conclusion, this means we have two separate issues:
PowerDNS should return the NSEC record even without the need for the `+dnssec` flag to be set. (but this is not related to the Unbound issue)
**AND**
Unbound should return the NSEC record when asked for it. Even if all above is PowerDNS's fault, this still happens even when querying the authoritative server of your company with Unbound.

Please consider re-opening the issue and fixing it.

**Comment by wcawijngaards** [comment]
You are trying to query for the type NSEC and that is not really supported. Also for type NSEC3 where the spec even disallows type NSEC3 queries. What seems to happen is that the CNAME is in cache. Unbound picks up the cname and then continues to create an answer from there. That would always happen. Apart from that, if Unbound queries the authoritative server, if that server responds with a CNAME unbound chases it, but if that server responds with the NSEC answer, then Unbound is likely going to create an answer out of that.

The issue is the data next to the CNAME. CNAME is defined to redirect all queries to the target, and this is also for other types, as the user wants to know what the data item at the target name is. And then needs the DNSSEC NSEC information for that target name to find out in the type bitmap if that type does not exist.

I think your query for type NSEC is not going to work. Try querying for another type, what did you really want? If you want to download the zone itself, ask for that, and use AXFR. If you want to have the type bitmap, ask for a nonexistent type. If you want a particular type ask for that.

It is interesting that servers differ in behaviour. But here unbound is just modifying the answer by what is in cache and the behaviour from the authoritative server. Why do you think PowerDNS is involved? It would be nicer if you got consistent answers from everyone.

**Comment by firstdorsal** [comment]
PowerDNS is my authoritative server, for that it is involved.
I used Cloudflare and Googles public DNS before to perform zone enumeration and wanted to switch to unbound for privacy reasons.
I'm just trying to query unbound for NSEC records in the first step (walking the zone) but can in no way get an answer from unbound, related to the requested NSEC record, in the case of an existing CNAME record.

**Comment by firstdorsal** [comment]
I guess i found a solution to my problem.
I'm using bind now...
It works as expected in this case.
